libobs/obs.c

```diff
@@ -2563,20 +2564,31 @@ obs_data_t *obs_save_source(obs_source_t *source)
 		obs_transition_save(source, source_data);
 
 	pthread_mutex_lock(&source->filter_mutex);
+	da_init(filters_copy);
+	da_reserve(filters_copy, source->filters.num);
 
-	if (source->filters.num) {
-		for (size_t i = source->filters.num; i > 0; i--) {
-			obs_source_t *filter = source->filters.array[i - 1];
+	for (size_t i = 0; i < source->filters.num; i++) {
+		obs_source_t *filter =
+			obs_source_get_ref(source->filters.array[i]);
+		if (filter)
+			da_push_back(filters_copy, &filter);
+	}
+
+	pthread_mutex_unlock(&source->filter_mutex);
+
+	if (filters_copy.num) {
+		for (size_t i = filters_copy.num; i > 0; i--) {
+			obs_source_t *filter = filters_copy.array[i - 1];
 			obs_data_t *filter_data = obs_save_source(filter);
 			obs_data_array_push_back(filters, filter_data);
 			obs_data_release(filter_data);
+			obs_source_release(filter);
 		}
 
 		obs_data_set_array(source_data, "filters", filters);
+		da_free(filters_copy);
```

Comment:
This is potentially a memory leak, as `filters_copy` was allocated with `source->filters.num` and not `filters_copy.num`.

Reply:
I suppose we should either remove the call to `da_reserve()` or make the call to `da_free()` unconditionally.

Reply:
Ah woops, you're right. My apologies for that. Will fix.

Reply:
Fixed in e35824452130320a281f3ed22fe68505fbaf0687 -- sorry for the trouble, and thanks for spotting it.